### ISSUE TITLE
Replacing a value with null across many columns is very slow in a lazyframe

### ISSUE THREAD

**Issue by mmcdermott** [pola-rs/polars#10887]
I don't have full details on this issue or a simple code example, but I wanted to report the following issue I've observed. I have a somewhat small dataframe that has a number of "count" columns which are often dominated by zeros. I have many of these dataframes (so in total, it is a lot of data), and I've observed using fully collected data-frames that replacing zeros by nulls can significantly speed up read/write time from disk. My code that produces these dataframes, however, uses lazy frames. I added this block of code to do the replacement to my lazy frame, just before it was collected and written to disk:

```python
df = df.with_columns(pl.when(cs.ends_with('count') != 0).then(cs.ends_with('count')).keep_name())
```

And suddenly my code slowed down dramatically. In comparison, If I instead `collect` the dataframe before running that line, then the overall code speeds up (because, as noted previously) the final writes on the dataframes with Nulls are faster. This seems like an error in the LazyFrame collection system, so I wanted to highlight it.

You can see the code I'm talking about here for an example here: https://github.com/mmcdermott/EventStreamGPT/blob/7e8cdda14bc8cea4674d14d7106c1cc6c8d15e47/EventStream/data/dataset_polars.py#L1766

Unfortunately I haven't constructed a minimum working example of the problem yet, but time permitting I'll try to add one soon. Regardless, I thought I should highlight this issue in case others experience the same thing.

**Comment by ritchie46** [comment]
Can you make a minimal example? Then we can do some profiling.

**Comment by mmcdermott** [comment]
Yes, I will try, but it may take me some time. Apologies for the delay!

**Comment by daviskirk** [comment]
@mmcdermott did you find a solution here?
I'm seeing the exact same thing.  Can't come up with a simple example though. 

However, I don't think it is inherently related to ANY lazyframe. Doing a collect and jumping back to the lazy API works ok as well, so my guess is that we're repeating operations from the lazyframes operation history or not correctly eliminating sub-expressions.

fast:

`lazy_df.collect().with_columns(pl.when(...).then(...))`
`lazy_df.collect().lazy().with_columns(pl.when(...).then(...))`

slow:

`lazy_df.with_columns(pl.when(...).then(...))`

**Comment by mmcdermott** [comment]
To be honest I forgot about this issue and don't recall if I ever really
solved it or if I just converted this operation back to operating over Data
Frames. Sorry!

**Comment by cBournhonesque** [comment]
This is still an issue.
```
import polars as pl
import numpy as np
import polars.selectors as cs

cols = {f"{i}": [1.0, np.inf, None, np.nan] for i in range(100)}
a = pl.LazyFrame(cols)
b = a.select(pl.when(cs.float().is_finite()).then(cs.float()).otherwise(3.0))
c = a.select(cs.float().replace([np.inf, -np.inf], 3.0).fill_nan(3.0).fill_null(3.0))
```

```
%timeit b.collect(engine="streaming")
1.89 ms ± 24.2 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

```
%timeit c.collect(engine="streaming")
310 ms ± 3.99 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

```
%timeit b.collect(engine="in-memory")
1.44 ms ± 22.7 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

```
%timeit c.collect(engine="in-memory")
7.2 ms ± 44 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
